### ISSUE TITLE
Create smoke test to prevent start up crashes when servers return unsuccessful requests

### ISSUE THREAD

**Issue by mcomella** [mozilla-mobile/fenix#18041]
This is very similar to https://github.com/mozilla-mobile/fenix/issues/18040. The distinction is that that bug is for no network access; **this bug is for network access but bad results.**

We experienced a start up crash when a server had an error in the response https://github.com/mozilla-mobile/fenix/issues/17762. I don't believe there was any generic smoke test put in place to prevent this class of errors (crashes due to 1) non-successful server requests or 2) servers returning bad data). We should add such a test. I think we can only automate this test by creating a `MockWebServer` that returns bad data.

To be explicit, **this is not something the performance team intends to work on** in the short term but **it's important to us** because crashes like these cause our performance tests in mozilla-central to fail, which disrupts our engineers' work stream as we investigate such crashes and possibly missing the cause of regressions as we lose performance test data.

┆Issue is synchronized with this [Jira Task](https://mozilla-hub.atlassian.net/browse/FNXV2-9740)

**Comment by Mugurell** [comment]
I was thinking that we could have a custom runner that would start a mockWebServer in Application's onCreate().
So that it would be a wrapper over the entire app functionality that would catch all network calls not yet intercepted by other mockWebServer instances below - that could respond more appropriately for the particular tests.
But if that is going to intercept all calls to the outside it would probably create other issues and greately increase the maintenance costs since it will probably still need to differentiate between different endpoints and respond with different data.
And then we get to the requirements here - the same api call from the app should receive different error responses. Maybe one random error response from a bigger pool of such responses.
So maybe we could have that encompassing mockWebServer with a pool of responses (with both valid / error response?) that would be returned randomly?

Another scenario that I don't know if it's possible or even worth the effort is adding a new lint rule that would identify a method call for network communication is not tested. This would probably not solve much though...
I was thinking at this because I'd say this "unexpected response" scenarios could very easily be unit tested when the features are implemented.

If going with the mockWebServer scenario I described above we would still be unable to test the same network call with more return values. We could only send one response from a big pool of success / error responses so this would not offer great confidence. But with more care for how network calls are used and more unit tests for these methods potential issues should be reduced to a minimum.

@sv-ohorvath @mcomella Curious what are your thoughts.

**Comment by sv-ohorvath** [comment]
So far, from what we've discussed, this is definitely more suitable for unit tests (sounds a lot easier and safer for sure).

**Comment by mcomella** [comment]
> this is definitely more suitable for unit tests (sounds a lot easier and safer for sure).

In theory, I agree – I think the unit tests should focus on robustly testing the error handling. That being said, in practice I think it's easy to forget to write a test for an important error case that causes a start up crash in Nightly. I filed this bug to try to solve this second problem, the start up crashes, which I think might be best served by an end-to-end test (i.e. actually starting the app and seeing if it crashes under some condition).

Unfortunately, I'm not familiar with the mockWebServer API to suggest specific implementation details but I'll do my best.

> So that it would be a wrapper over the entire app functionality that would catch all network calls not yet intercepted by other mockWebServer instances below - that could respond more appropriately for the particular tests.

To clarify, do you want this capturing to run during all integration tests? I'd probably prefer to isolate it to this specific test – as you mention, adding a mockWebServer to intercept all network requests adds complexity and maintenance costs. I think, for this specific test, we'd make all outgoing requests have unsuccessful responses.

> the same api call from the app should receive different error responses. Maybe one random error response from a bigger pool of such responses.

This is a really good point – if we don't test the wide range of error responses, we might experience a start up crash and this test wouldn't do anything. I think randomness in tests is a code smell though: it can create intermittent test failures that are difficult to debug. In the ideal case, we'd test the entire large range of error responses on start up. However, at least in my proposal to start the app for each test, it'd be impractical because of the time these tests would take to run. We should probably leave the robust handling for unit tests.

As such, I think we should start with the simple thing – create maybe one or two test cases for this scenario. Perhaps "success code is error (e.g. 404)" and "nonsensical data returned (e.g. an invalid JSON string)". If we experience the start up crash again after writing this test, we can iterate and improve the test (or give up if it's impractical to fix). If we don't experience it again, maybe this test is working! :)

**To summarize,** I'm thinking of this as a really simple smoke test to catch start up crashes caused by devs not properly handling bad network responses because they forgot to write the right unit tests for it.

> adding a new lint rule

I like the thinking on alternatives but I think this one is hard: network calls can be wrapped in other functions that need to be tested and lint rules aren't great at handling call hierarchies.

---

>  same api call from the app should receive different error responses.

It'd be interesting to have some capability that helps you write network unit tests by providing a large range of error responses, huh? Maybe too "perfect is the enemy of the good" though. :)

**Comment by sv-ohorvath** [comment]
Waiting for https://github.com/mozilla-mobile/fenix/issues/18385

**Comment by mcomella** [comment]
> crashes due to 1) non-successful server requests or 2) servers returning bad data).

AIUI, we addressed case 1 but not case 2.

@sv-ohorvath It looks like #18385 was closed – do we still plan to work on this issue? If not, I think we can close it – case 2 seems like a nice-to-have but it seems less problematic than case 1.

**Comment by sv-ohorvath** [comment]
@mcomella I lost these between all the Github notifications I suppose. I didn't abandon yet the work on this, the original work is still on a branch waiting for #18385, but I don't know who will have the time to pick it up or when...
